sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -15,190 +14,347 @@
 import com.azure.cosmos.implementation.RetryPolicyWithDiagnostics;
 import com.azure.cosmos.implementation.RetryWithException;
 import com.azure.cosmos.implementation.RxDocumentServiceRequest;
-import com.azure.cosmos.implementation.apachecommons.lang.time.StopWatch;
+import com.azure.cosmos.implementation.apachecommons.lang.tuple.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import reactor.core.publisher.Mono;
 
 import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Supplier;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class GoneAndRetryWithRetryPolicy
+    extends RetryPolicyWithDiagnostics
+    implements LastRetryWithExceptionHolder, LastRetryWithExceptionProvider {
```

Comment:
why do we need these two new interfaces `LastRetryWithExceptionHolder, LastRetryWithExceptionProvider`? 

Is it to provide the linkage between the inner policies and the outer policy?

The inner class should be able to access the outer class field. RetryWithPolicy should be able to access  the `GoneAndRetryWithRetryPolicy.retryWithException` with `GoneAndRetryWithRetryPolicy.this.retryWithException`
https://www.programiz.com/java-programming/nested-inner-class

I think we should be able to get rid of the two new interfaces with above appraoch.

Reply:
Yeah - wasn't aware of the <Class>.this.xxx notion